firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp

```diff
@@ -33,214 +33,93 @@ namespace Pufferfish::Driver::Serial::Nonin {
-
-  /**
-   * spo2 Fast: Non-slew limited saturation with 4-beat averaging in standard
-   * mode. sensorData 10th frame of packet contains spo2-D-Bit6 t0 spo2-D-Bit0
-   * in bit6 to bit0
-   */
-  sensor_measurements.spo2_d_fast = get_6bit_data(packet_data[spo2_d_fast_index][byte3]);
-
-  /**
-   * spo2 B-B: Un-averaged, non-slew limited, beat to beat value in standard
-   * mode sensorData 11th frame of packet contains spo2-D-Bit6 t0 spo2-D-Bit0 in
-   * bit6 to bit0
-   */
-  sensor_measurements.spo2_d_beat = get_6bit_data(packet_data[spo2_d_beat_index][byte3]);
-
-  /**
-   * E-Heart Rate: 8-beat average values in standard mode.
-   * E-HR MSB : sensorData 14th frame of packet contains HR-Bit8 and HR-Bit7 in
-   * bit1 and bit0 E-HR LSB : sensorData 15th frame of packet contains HR-Bit6
-   * t0 HR-Bit0 in bit6 to bit0
-   */
-  sensor_measurements.e_heart_rate = get_9bit_data(
-      packet_data[e_heart_rate_index][byte3], packet_data[e_heart_rate_index + 1][byte3]);
-
-  /**
-   * E-spo2: 8-beat average values in standard mode.
-   * sensorData 16th frame of packet contains spo2-Bit6 t0 spo2-Bit0 in bit6 to
-   * bit0
-   */
-  sensor_measurements.e_spo2 = get_6bit_data(packet_data[e_spo2_index][byte3]);
-
-  /**
-   * E-spo2-D: 8-beat average displayed values in display mode
-   * sensorData 17th frame of packet contains spo2-D-Bit6 to E-spo2-D-Bit0 in
-   * bit6 to bit0
-   */
-  sensor_measurements.e_spo2_d = get_6bit_data(packet_data[e_spo2_d_index][byte3]);
-
-  /**
-   * HR-D: 8-beat average values in standard mode.
-   * HR-D MSB : sensorData 20th frame of packet contains HR-Bit8 and HR-Bit7 in
-   * bit1 and bit0 HR-D LSB : sensorData 21st frame of packet contains HR-Bit6
-   * t0 HR-Bit0 in bit6 to bit0
-   */
-  sensor_measurements.heart_rate_d = get_9bit_data(
-      packet_data[heart_rate_d_index][byte3], packet_data[heart_rate_d_index + 1][byte3]);
-
-  /**
-   * E-HR-D: 8-beat average displayed values in display mode
-   * E-HR-D MSB : sensorData 22nd frame of packet contains HR-Bit8 and HR-Bit7
-   * in bit1 and bit0 E-HR-D LSB : sensorData 23rd frame of packet contains
-   * HR-Bit6 t0 HR-Bit0 in bit6 to bit0
-   */
-  sensor_measurements.e_heart_rate_d = get_9bit_data(
-      packet_data[e_heart_rate_d_index][byte3], packet_data[e_heart_rate_d_index + 1][byte3]);
-
-  /* Update the PLETH and Status byte errors into the packet measurements */
-  for (frame_index = 0; frame_index < packet_size; frame_index++) {
+
+  // SpO2 data
+  sensor_measurements.spo2 = get_spo2_data(packet_data[spo2_index][byte3]);
+  sensor_measurements.spo2_d = get_spo2_data(packet_data[spo2_d_index][byte3]);
+  sensor_measurements.spo2_fast = get_spo2_data(packet_data[spo2_fast_index][byte3]);
+  sensor_measurements.spo2_b_b = get_spo2_data(packet_data[spo2_b_b_index][byte3]);
+  sensor_measurements.e_spo2 = get_spo2_data(packet_data[e_spo2_index][byte3]);
+  sensor_measurements.e_spo2_d = get_spo2_data(packet_data[e_spo2_d_index][byte3]);
```

Comment:
[for a future PR] I wonder what `byte3 means` and why we use it so many times. Maybe we can find a more descriptive name, e.g. have a `ByteIndices` namespace in Types.h with constants for `header_byte`, `status_byte`, `pleth_byte`, `payload_byte`/`data_byte`, and `checksum_byte`.